exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/okhttp/OkHttpRetrier.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.exporter.otlp.internal.okhttp;
+
+import io.opentelemetry.exporter.otlp.internal.RetryPolicy;
+import java.io.IOException;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import okhttp3.Interceptor;
+import okhttp3.Response;
+
+/** Retrier of OkHttp requests. */
+public final class OkHttpRetrier implements Interceptor {
+
+  private final RetryPolicy retryPolicy;
+  private final Function<Response, Boolean> isRetriableError;
+  private final Sleeper sleeper;
+  private final BoundedLongGenerator randomLong;
+
+  /** Constructs a new retrier. */
+  public OkHttpRetrier(RetryPolicy retryPolicy, Function<Response, Boolean> isRetriableError) {
+    this(
+        retryPolicy,
+        isRetriableError,
+        TimeUnit.NANOSECONDS::sleep,
+        bound -> ThreadLocalRandom.current().nextLong(bound));
+  }
+
+  // Visible for testing
+  OkHttpRetrier(
+      RetryPolicy retryPolicy,
+      Function<Response, Boolean> isRetriableError,
+      Sleeper sleeper,
+      BoundedLongGenerator randomLong) {
+    this.retryPolicy = retryPolicy;
+    this.isRetriableError = isRetriableError;
+    this.sleeper = sleeper;
+    this.randomLong = randomLong;
+  }
+
+  @Override
+  public Response intercept(Chain chain) throws IOException {
+    Response response = chain.proceed(chain.request());
+    if (!Boolean.TRUE.equals(isRetriableError.apply(response))) {
+      return response;
+    }
+    long nextBackoffNanos = retryPolicy.getInitialBackoff().toNanos();
+    // Made an attempt already, starting with the 2nd.
+    for (int attempt = 2; attempt <= retryPolicy.getMaxAttempts(); attempt++) {
+      // https://github.com/grpc/proposal/blob/master/A6-client-retries.md#exponential-backoff
+      // Careful - the document says "retry attempt" but our maxAttempts includes the original
+      // request. The numbers are tricky but the unit test should be more clear.
+      long upperBoundNanos = Math.min(nextBackoffNanos, retryPolicy.getMaxBackoff().toNanos());
+      long backoffNanos = randomLong.get(upperBoundNanos);
+
+      try {
+        sleeper.sleep(backoffNanos);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        return response;
```

Comment:
Do we want to return here or raise a runtime exception?

Reply:
From what I understand calling interrupt and returning is the correct way of handling an interrupt exception, though I've never really run into it differing much from throwing in practice

Reply:
Yeah I think in practice the thread will only be interrupted at shutdown so it shouldn't matter much that we return.